pkg/scheduler/flavorassigner/flavorassigner.go

```diff
@@ -218,33 +220,67 @@ func (m FlavorAssignmentMode) String() string {
 }
 
 type FlavorAssignment struct {
-	Name   kueue.ResourceFlavorReference
-	Mode   FlavorAssignmentMode
-	borrow int64
+	Name                    kueue.ResourceFlavorReference
+	Mode                    FlavorAssignmentMode
+	LastScheduleFlavorIndex int
+	borrow                  int64
 }
 
 // AssignFlavors assigns flavors for each of the resources requested in each pod set.
 // The result for each pod set is accompanied with reasons why the flavor can't
 // be assigned immediately. Each assigned flavor is accompanied with a
 // FlavorAssignmentMode.
 func AssignFlavors(log logr.Logger, wl *workload.Info, resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor, cq *cache.ClusterQueue, counts []int32) Assignment {
+	if wl.LastSchedule != nil {
+		if cq.Cohort == nil && !wl.LastSchedule.Equal(resourceFlavors, workload.FlavorResourceQuantities(cq.Usage), make(workload.FlavorResourceQuantities)) {
+			wl.LastSchedule = nil
+		}
+		if cq.Cohort != nil && !wl.LastSchedule.Equal(resourceFlavors, workload.FlavorResourceQuantities(cq.Usage), workload.FlavorResourceQuantities(cq.Cohort.Usage)) {
+			wl.LastSchedule = nil
+		}
+	}
+
 	if len(counts) == 0 {
-		return assignFlavors(log, wl.TotalRequests, wl.Obj.Spec.PodSets, resourceFlavors, cq)
+		return assignFlavors(log, wl.TotalRequests, wl.Obj.Spec.PodSets, resourceFlavors, cq, wl.LastSchedule)
 	}
 
 	currentResources := make([]workload.PodSetResources, len(wl.TotalRequests))
 	for i := range wl.TotalRequests {
 		currentResources[i] = *wl.TotalRequests[i].ScaledTo(counts[i])
 	}
-	return assignFlavors(log, currentResources, wl.Obj.Spec.PodSets, resourceFlavors, cq)
+	return assignFlavors(log, currentResources, wl.Obj.Spec.PodSets, resourceFlavors, cq, wl.LastSchedule)
 }
 
-func assignFlavors(log logr.Logger, requests []workload.PodSetResources, podSets []kueue.PodSet, resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor, cq *cache.ClusterQueue) Assignment {
+func assignFlavors(log logr.Logger, requests []workload.PodSetResources, podSets []kueue.PodSet, resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor, cq *cache.ClusterQueue, lastSchedule *workload.LastScheduleClusterQueueState) Assignment {
 	assignment := Assignment{
 		TotalBorrow: make(cache.FlavorResourceQuantities),
 		PodSets:     make([]PodSetAssignment, 0, len(requests)),
-		usage:       make(cache.FlavorResourceQuantities),
+		ScheduleState: workload.LastScheduleClusterQueueState{
```

Comment:
if there is a previous state, we already verified that it matches the previous state of the CQ and cohort. So in that case, we shouldn't need to create new maps, we can reuse the existing ones.

Reply:
done